text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+Dependency autotracking requires that tracked getters access their dependencies
+synchronously. Any access that happens asynchronously will not be detected as a
+dependency.
+
+This is most commonly encountered when trying to return a `Promise` from a
+tracked getter. Here's an example that would "work" but would never update if
+`firstName` or `lastName` change:
+
+```js
+class Person {
+  @tracked firstName;
+  @tracked lastName;
+
+  @tracked
+  get fullNameAsync() {
+    return this.reloadUser().then(() => {
+      return `${this.firstName} ${this.lastName}`;
+    });
+  }
+
+  async reloadUser() {
+    const response = await fetch('https://example.com/user.json');
+    const { firstName, lastName } = await response.json();
+    this.firstName = firstName;
+    this.lastName = lastName;
+  }
+
+  setFirstName(firstName) {
+    // This should cause `fullNameAsync` to update, but doesn't, because
+    // firstName was not detected as a dependency.
+    this.firstName = firstName;
+  }
+}
+```
+
+One way you could address this is to ensure that any dependencies are consumed
+synchronously:
+
+```js
+@tracked
+get fullNameAsync() {
+  // Consume firstName and lastName so they are detected as dependencies.
+  let { firstName, lastName } = this;
+
+  return this.reloadUser().then(() => {
+    // Fetch firstName and lastName again now that they may have been updated
+    let { firstName, lastName } = this;
+    return `${this.firstName} ${this.lastName}`;
+  });
+}
+```
+
+However, **modeling async behavior as tracked properties is an incoherent
+approach and should be discouraged**. Tracked properties are intended to hold
+simple state, or to derive state from data that is available synchronously.
+
+But asynchrony is a fact of life in web applications, so how should we deal with
+async data fetching?
+
+**In keeping with Data Down, Actions Up, async behavior should be modeled as
+methods that set tracked properties once the behavior is complete.**
+
+Async behavior should be explicit, not a side-effect of property access. Today's
+computed properties that rely on caching to only perform async behavior when a
+dependency changes are effectively reintroducing observers into the programming
+model via a side channel.
+
+A better approach is to call a method to perform the async data fetching, then
+set one or more tracked properties once the data has loaded. We can refactor the
+above example back to a synchronous `fullName` tracked property:
+
+```js
+class Person {
+  @tracked firstName;
+  @tracked lastName;
+
+  @tracked
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+
+  async reloadUser() {
+    const response = await fetch('https://example.com/user.json');
+    const { firstName, lastName } = await response.json();
+    this.firstName = firstName;
+    this.lastName = lastName;
+  }
+}
+```
+
+Now, `reloadUser()` must be called explicitly, rather than being run implicitly
+as a side-effect of consuming `fullName`.
```

Comment:
we should consider added a dedicated lint rule to eslint-plugin-ember that detects `.then()` and `await` in tracked property getters

Reply:
@Turbo87 Strong +1 to this suggestion.

Reply:
thinking about it, we should probably have the same for regular computed properties too... 🤔

Reply:
there _is_ a use case for this type of behavior though: https://github.com/NullVoxPopuli/ember-computed-promise-monitor

async relationships, for example.
(though, I'm currently of the opinion that they should be avoided at all costs, due to their complexity in accessing the properties on the async relationships)